### PR TITLE
OSOE-150: ⭐ Add PowerShell static code analysis to Utility-Scripts

### PR DESCRIPTION
**Pull request by sarahelsaig**
OSOE-150

### CODE REVIEW

**Review by Piedone — APPROVED** [review]

**Comment by Piedone** [comment]
Does it really not find any other issue in all of the other scripts?

**Comment by sarahelsaig** [comment]
Argh, I've just realized! Most of the files are psm1, not ps1, so they aren't found by the recursive search.

**Comment by Piedone** [comment]
I see then. Otherwise congrats @BenedekFarkas :D.

**Comment by BenedekFarkas** [comment]
Congrats on what?

**Comment by Piedone** [comment]
On writing dozens of PS scripts with a single analyzer violation!

**Comment by sarahelsaig** [comment]
Sadly no, but about ⅔ of them have no violations indeed. So congrats are still in order.
image

**Comment by BenedekFarkas** [comment]
Thanks guys! :D

File: Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1 [hunk 1]
```diff
@@ -16,29 +16,33 @@ function Reset-AppDataFolder
...
+        [string] $Path,
 
         [switch] $KeepLogFiles
     )
 
     Process
-    {        
-        if (!(Test-Path $Path))
+    {
+        if (-not (Test-Path $Path))
```
Comment (Piedone) [inline Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1:31]: Why is `-not` and `-isnot` better?

File: OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1 [hunk 2]
```diff
@@ -1,10 +1,10 @@
 ﻿<#
 .Synopsis
-   Initializes an Orchard Core solution for a git repository. Deprecated, use Init-OrchardCoreSolution instead.
+   Initializes an Orchard Core solution for a git repository. Deprecated, use Initialize-OrchardCoreSolution instead.
 #>
 
 
-function Init-OrchardCore
+function Initialize-OrchardCore
```
Comment (Piedone) [inline OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1:7]: Keep this, since this is a proxy for an old name.

File: OrchardCore/Initialize-OrchardCoreSolution/Initialize-OrchardCoreSolution.psm1 [hunk 3]
```diff
@@ -6,11 +6,11 @@
    Initializes an Orchard Core solution using the latest released Orchard Core NuGet packages at the current location or under the given path, and adds a suitable .gitignore file. Optionally creates an initial module and/or theme, and optionally uses a given NuGet source.
 
 .EXAMPLE
-   Init-OrchardCoreSolution -Name "FancyWebsite" -Path "D:\Work\FancyWebsite" -ModuleName "FancyWebsite.Core" -ThemeName "FancyWebsite.Theme" -NuGetSource "https://nuget.cloudsmith.io/orchardcore/preview/v3/index.json"
+   Initialize-OrchardCoreSolution -Name "FancyWebsite" -Path "D:\Work\FancyWebsite" -ModuleName "FancyWebsite.Core" -ThemeName "FancyWebsite.Theme" -NuGetSource "https://nuget.cloudsmith.io/orchardcore/preview/v3/index.json"
```
Comment (Piedone) [inline OrchardCore/Initialize-OrchardCoreSolution/Initialize-OrchardCoreSolution.psm1:9]: Add a proxy script for backwards compatibility, since we our Dojo Course refers to this script.

File: SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1 [hunk 4]
```diff
@@ -15,23 +15,24 @@ function Get-DefaultSqlServerName
...
     Process
     {
-        $serverServices = (Get-WmiObject win32_Service -Computer $env:COMPUTERNAME | Where-Object { $PSItem.Name -match "MSSQL" -and $PSItem.PathName -match "sqlservr.exe" })
+        $serverServices = Get-CimInstance win32_Service |
+            Where-Object { $_.Name -match "MSSQL" -and $_.PathName -match "sqlservr.exe" }
         $servicePath = ""
 
         # No SQL Servers installed.
-        if ($serverServices -eq $null)
+        if ($null -eq $serverServices)
```
Comment (Piedone) [inline SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1:23]: Why the reverse order?

File: SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1 [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<#
+﻿<#
```
Comment (Piedone) [inline SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1:1]: [Using this script from the contest menu](https://github.com/Lombiq/Utility-Scripts/#:~:text=To%20be%20able,from%20here.) now yields this:
image

File: Utilities/Reload-Module/Reload-Module.psm1 [hunk 6]
```diff
@@ -11,9 +11,11 @@
     Reload-Module "C:\Path\To\PSModules"
 #>
 
-
 function Reload-Module
 {
+    [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute(
+        'PSUseApprovedVerbs',
+        Justification = 'Use distinctive name to confusion with existing cmdlets such as Import-Module.')]
```
Comment (Piedone) [inline Utilities/Reload-Module/Reload-Module.psm1]:
```suggestion
        Justification = 'Use distinctive name to avoid confusion with existing cmdlets such as Import-Module.')]
```

File: Utilities/Stop-Processes/Stop-Processes.psm1 [hunk 7]
```diff
@@ -1,102 +0,0 @@
-<#
```
Comment (Piedone) [inline Utilities/Stop-Processes/Stop-Processes.psm1:1]: Why was this deleted?

**Review by Piedone — CHANGES_REQUESTED** [review]
Merge from dev.

**Comment by Piedone** [comment]
@BenedekFarkas FYI, see the changes. A couple of script renames, mostly ones with plural names. Do you want to review the changes?

File: Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1 (hunk 1 above)
Comment (sarahelsaig) [reply Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1:31]: They are the canonical comparison operators. `!` is not mentioned [in the official documentation](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_comparison_operators?view=powershell-7.2) as an operator at all. So to me `!` feels like an alias to `-not` (except it's worse becuase afaik it's undocumented).

The same applies to `-isnot`, and also it makes things simpler because you get the result with a single operator instead of the combination of a prefix and an infix operator. Since the operators are so bulky, I'm pretty happy that MS gave us `-isnot`, `-notlike`, `-notmatch`, etc.

File: OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1 (hunk 2 above)
Comment (sarahelsaig) [reply OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1:7]: Should I rename the file, or just add `[alias("Init-OrchardCore")]`?

File: SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1 (hunk 4 above)
Comment (sarahelsaig) [reply SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1:23]: Otherwise we get this warning:
> `$null` should be on the left side of equality comparisons.
> PSScriptAnalyzer([PSPossibleIncorrectComparisonWithNull](https://docs.microsoft.com/en-us/powershell/utility-modules/psscriptanalyzer/rules/possibleincorrectcomparisonwithnull?view=ps-modules))

File: Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1 (hunk 1 above)
Comment (Piedone) [reply Orchard1/Reset-AppDataFolder/Reset-AppDataFolder.psm1:31]: I see, thanks.

File: OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1 (hunk 2 above)
Comment (Piedone) [reply OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1:7]: Alias is OK too.

File: OrchardCore/Initialize-OrchardCoreSolution/Initialize-OrchardCoreSolution.psm1 (hunk 3 above)
Comment (Piedone) [reply OrchardCore/Initialize-OrchardCoreSolution/Initialize-OrchardCoreSolution.psm1:9]: Or an alias.

File: SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1 (hunk 4 above)
Comment (Piedone) [reply SqlServer/Get-DefaultSqlServerName/Get-DefaultSqlServerName.psm1:23]: Interesting.

File: SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1 (hunk 5 above)
Comment (sarahelsaig) [reply SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1:1]: My bad, looks like the behavior of `Join-Path` is just too different between `pwsh` and `powershell`. I replaced it with `[System.IO.Path]::Combine`, that at least works the same across versions.

File: Utilities/Reload-Module/Reload-Module.psm1 (hunk 6 above)
Comment (sarahelsaig) [reply Utilities/Reload-Module/Reload-Module.psm1]: Assuming we want to _avoid_ confusion
image

File: Utilities/Stop-Processes/Stop-Processes.psm1 (hunk 7 above)
Comment (sarahelsaig) [reply Utilities/Stop-Processes/Stop-Processes.psm1:1]: > `Stop-Process` is a cmdlet that is included with PowerShell (version core-6.1.0-windows) whose definition should not be overridden
> PSScriptAnalyzer([PSAvoidOverwritingBuiltInCmdlets](https://docs.microsoft.com/en-us/powershell/utility-modules/psscriptanalyzer/rules/avoidoverwritingbuiltincmdlets?view=ps-modules))

There is already a `Stop-Process` cmdlet in both PS5 and PS7. The only way `Stop-Process` is used in other scripts is by passing it a single process ID or a `Process` object. Both are supported by the built-in cmdlet so this one is not needed.

File: OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1 (hunk 2 above)
Comment (BenedekFarkas) [reply OrchardCore/Initialize-OrchardCore/Initialize-OrchardCore.psm1:7]: Yep, `Initialize` is the approved verb and `Init` is not recommended, but keeping it for backwards-compatibility is fine. Same for `Initialize-OrchardCoreSolution`.
https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands

File: Utilities/Stop-Processes/Stop-Processes.psm1 (hunk 7 above)
Comment (BenedekFarkas) [reply Utilities/Stop-Processes/Stop-Processes.psm1:1]: That's right, but this is `Stop-Processes`, as in multiple and that's the whole point - it allows killing multiple processes by name with a single command. We use it in TeamCity to clean up lingering processes after UI test execution.

**Comment by BenedekFarkas** [comment]
> @BenedekFarkas FYI, see the changes. A couple of script renames, mostly ones with plural names. Do you want to review the changes?

Yep!

File: DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1 [hunk 8]
```diff
@@ -1,4 +1,4 @@
-function Get-VisualStudioProjectNuGetPackages
+﻿function Get-VisualStudioProjectNuGetPackage
```
Comment (BenedekFarkas) [inline DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1]: Changing to singular is to conform to a convention/analyzer? I'm asking because making this plural has a meaning, i.e. you receive all the NuGet packages referenced by the project, so potentially you're not just getting a single entry of something.

Also, looking at the singular vs. plural name without context evokes a completely different idea (at least to me) about the script's behavior.

File: AddPathToPSModulePath.ps1 [hunk 9]
```diff
@@ -22,7 +22,7 @@ $paths = [Environment]::GetEnvironmentVariable("PSModulePath", "Machine").Split(
 if(!$paths.Contains($Path))
 {
     [System.Environment]::SetEnvironmentVariable("PSModulePath", [string]::Join(";", $paths + $Path), "Machine")
-    Write-Information "The path `"$Path`" was successfully added to the PSModulePath environment variable."
+    Write-Verbose "The path `"$Path`" was successfully added to the PSModulePath environment variable."
```
Comment (BenedekFarkas) [inline AddPathToPSModulePath.ps1:25]: Does `Write-Information` have a similar downside as `Write-Host`? I'd argue that in this case (and some other ones below) displaying a message is useful (so it should be `Write-Output`), even without `-Verbose` so that the user knows what's going on.

Personally I absolutely hate that printing nothing to the console means successful execution. It goes against UX and safety/security in so many other areas too.

File: DotNetCLI/Get-VisualStudioSolutionProjectPaths/Get-VisualStudioSolutionProjectPaths.psm1 [hunk 10]
```diff
@@ -1,4 +1,4 @@
-function Get-VisualStudioSolutionProjectPaths
+﻿function Get-VisualStudioSolutionProjectPath
```
Comment (BenedekFarkas) [inline DotNetCLI/Get-VisualStudioSolutionProjectPaths/Get-VisualStudioSolutionProjectPaths.psm1:1]: `Get-VisualStudioSolutionProjectsPath` (project becomes plural) instead? Signals the intention more clearly.

File: Utilities/Reload-Module/Reload-Module.psm1 [hunk 11]
```diff
@@ -27,24 +29,27 @@ function Reload-Module
     {
         if (Test-Path($Path) -PathType Container)
         {
-            $modules = Get-ChildItem -Path $Path -Recurse -File | Where-Object { [System.IO.Path]::GetExtension($PSItem.FullName).Equals(".psm1", [System.StringComparison]::InvariantCultureIgnoreCase) }
+            $modules = Get-ChildItem -Path $Path -Recurse -File |
+                Where-Object { [System.IO.Path]::GetExtension($PSItem.FullName) -match '^\.psm1$' }
```
Comment (BenedekFarkas) [inline Utilities/Reload-Module/Reload-Module.psm1]: Could/should this use `$_.FullName` instead?

File: Utilities/Rename-ChildItemsToAsciiRecursively/Rename-ChildItemsToAsciiRecursively.psm1 [hunk 12]
```diff
@@ -19,7 +19,7 @@ function Rename-ChildItemsToAsciiRecursively
 
             if ($item.Name -ne $asciiName)
             {
-                Write-Host ("Renaming `"$($item.FullName)`" to `"$asciiName`".")
+                Write-Verbose "Renaming `"$($item.FullName)`" to `"$asciiName`"."
```
Comment (BenedekFarkas) [inline Utilities/Rename-ChildItemsToAsciiRecursively/Rename-ChildItemsToAsciiRecursively.psm1]: Let's add a non-verbose message here about the number of items that were renamed.

File: Utilities/Set-FileContent/Set-FileContent.psm1 [hunk 13]
```diff
@@ -35,6 +35,6 @@ function Set-FileContent
 
         (Get-Content $FilePath) | Foreach-Object { $PSItem -replace $Match, $ReplaceWith } | Set-Content $FilePath
 
-        Write-Host ("Successfully replaced all occurrences of `"$Match`" with `"$ReplaceWith`" in the file `"$FilePath`"!")
+        Write-Verbose "Successfully replaced all occurrences of `"$Match`" with `"$ReplaceWith`" in the file `"$FilePath`"!"
```
Comment (BenedekFarkas) [inline Utilities/Set-FileContent/Set-FileContent.psm1:38]: This is also a case where successful execution would have no output whatsoever without `-Verbose`, IMO this should be Write-Output.

**Review by BenedekFarkas — CHANGES_REQUESTED** [review]

File: Utilities/Stop-Processes/Stop-Processes.psm1 (hunk 7 above)
Comment (BenedekFarkas) [reply Utilities/Stop-Processes/Stop-Processes.psm1:1]: Just checked the other changes - so instead of
`Stop-Processes -Name <list of process names>`
we should have
`Get-ProcessId -Name <list of process names> | Stop-Process`
right?

Btw I'd like to note that our original intention was to be able to kill a process:
1. By name, regardless of command line.
2. By name and command line.
3. By command line, regardless of name.

AFAIK 3. is not covered by the changes, because the name parameter is mandatory, but we don't actually need that in our TC builds anymore.

File: DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1 (hunk 8 above)
Comment (sarahelsaig) [reply DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1]: The rule is:
> The cmdlet 'Get-VisualStudioProjectNuGetPackages' uses a plural noun. A singular noun should be used instead.
> PSScriptAnalyzer([PSUseSingularNouns](https://docs.microsoft.com/en-us/powershell/utility-modules/psscriptanalyzer/rules/usesingularnouns?view=ps-modules))

In this case you are returning a package list for a project. This is the same type of activity as returning a list of files for a directory with `Get-ChildItem` and that's singular as well. As far as I can recall all of built-in the `[array]`-returning cmdlets are singular.

Comment (BenedekFarkas) [reply DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1]: Understood, thanks!

File: AddPathToPSModulePath.ps1 (hunk 9 above)
Comment (sarahelsaig) [reply AddPathToPSModulePath.ps1:25]: `Write-Information` outputs to a separate, object stream. It's rather for passing structured (and tagged) data. It's not problematic, like `Write-Host`, but it's not made for the purpose it's used here. `Write-Verbose` is the correct way to display status text. 

> Personally I absolutely hate that printing nothing to the console means successful execution.

Actually `$?` being `$True` is what means successful execution. :D It's extremely normal for scripts to not print anything on success, becuase you don't want to flood the log files with rubbish when they are executed in automation. Since scripts are meant to be automated, UX isn't really an issue and imo not leaking implementation information is a good security practice. I'm used to working with GNU utilities so this is the normal for me. [The dude who invented PowerShell agrees too. ](https://www.jsnover.com/blog/2013/12/07/write-host-considered-harmful/#:~:text=powershell%20is%20about%20automation.%20show%20comforting%20information%20is%20often%20helpful%20the%20first%20couple%20of%20times%20you%20run%20a%20script%20but%20then%20after%20that%2C%20it%20because%20an%20annoyance.)
You can always just call the cmdlet with `-Verbose` to retain the old behavior, and now all the status messages are nice and yellow so I think it's a win-win.

File: DotNetCLI/Get-VisualStudioSolutionProjectPaths/Get-VisualStudioSolutionProjectPaths.psm1 (hunk 10 above)
Comment (sarahelsaig) [reply DotNetCLI/Get-VisualStudioSolutionProjectPaths/Get-VisualStudioSolutionProjectPaths.psm1:1]: The same point applies here, the logic here is like an advanced `Get-ChildItem` as well.

(And now I start dreaming about a world here `Get-ChildItem` has a `-Type` parameter just like `New-Item`, and we could extend it with a custom `VisualStudioSolutionProject` type. That would be so neat...)

File: AddPathToPSModulePath.ps1 (hunk 9 above)
Comment (BenedekFarkas) [reply AddPathToPSModulePath.ps1:25]: That's a reasonable argument (I've read the article), but we should also consider that at least in Utility Scripts, many of these scripts are used to set up something by a user, not just for classic automation. But I agree with the sentiment about spamming the logs.

Please make a note in the Readme about the usage of the Verbose switch in case of an ambiguous or failed execution (i.e. that we conform to this convention about the usage of the Verbose switch).

File: Utilities/Rename-ChildItemsToAsciiRecursively/Rename-ChildItemsToAsciiRecursively.psm1 (hunk 12 above)
Comment (sarahelsaig) [reply Utilities/Rename-ChildItemsToAsciiRecursively/Rename-ChildItemsToAsciiRecursively.psm1]: I've rewritten the method to pipe out a new PSObject with the renames. (Also no need to manually recurse.) I think more fitting than a string message. After all, we like PS because we can output structured data and not just strings.
image

Comment (BenedekFarkas) [reply Utilities/Rename-ChildItemsToAsciiRecursively/Rename-ChildItemsToAsciiRecursively.psm1]: OK, thanks!

File: Utilities/Set-FileContent/Set-FileContent.psm1 (hunk 13 above)
Comment (sarahelsaig) [reply Utilities/Set-FileContent/Set-FileContent.psm1:38]: I think utility scripts is especially where you should only have verbose for status, because these are most likely to be invoked from a script instead of by a user.

File: Utilities/Stop-Processes/Stop-Processes.psm1 (hunk 7 above)
Comment (sarahelsaig) [reply Utilities/Stop-Processes/Stop-Processes.psm1:1]: The built-in `Stop-Process` already stops multiple processes. You can provide multiple program names, process IDs or `Process` objects to it.
image

For example this works on Windows Powershell (just verified):
```powershell
Stop-Process -Name node, bash, python
```

If you want to kill multiple processes by name and command line too, that would be:
```powershell
$nameAndCommandLine = @{ Name = "node" }, @{ Name = "dotnet"; CommandLine = "SomeApp" }
$nameAndCommandLine | % { Get-ProcessId @_ } | Stop-Process
```

File: Utilities/Set-FileContent/Set-FileContent.psm1 (hunk 13 above)
Comment (BenedekFarkas) [reply Utilities/Set-FileContent/Set-FileContent.psm1:38]: Hmm, I think I wanted to comment this on another file (which is more likely to be used manually), but given the other conversation and Readme update, I agree with your assessment.

File: Utilities/Stop-Processes/Stop-Processes.psm1 (hunk 7 above)
Comment (BenedekFarkas) [reply Utilities/Stop-Processes/Stop-Processes.psm1:1]: Oh, I see! The situation was a bit different back when the command line filter was a requirement, so that's why I factored this all out into `Stop-Processes` and I forgot that Stop-Process accepts multiple process names. That's enough for TeamCity now, actually. Thanks for explaining!

File: Utilities/Reload-Module/Reload-Module.psm1 [hunk 14]
```diff
@@ -13,9 +13,9 @@
 
 function Reload-Module
 {
-    [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute(
+    [Diagnostics.CodeAnalysis.SuppressMessage(
```
Comment (Piedone) [inline Utilities/Reload-Module/Reload-Module.psm1:16]: I get this when running this from the root script: 
image

File: SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1 (hunk 5 above)
Comment (Piedone) [reply SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1:1]: I still get this with the latest source.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1 (hunk 5 above)
Comment (sarahelsaig) [reply SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1:1]: You shouldn't get the exact same error, becuase the affected code was replaced with something else. But I've noticed, that the `20..12` range is super tiny, for me the exe is actually in the "160" directory. So I've rewritten that part from scratch to just use `Get-ChildItem` (and select the numeric directory names) instead of using a hard coded integer range. This worked for me for sure:
image

File: Utilities/Reload-Module/Reload-Module.psm1 (hunk 14 above)
Comment (sarahelsaig) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: Turns out that second empty parameter in SuppressMessageAttribute is necessary after all (but only in PS 5) :(

Comment (Piedone) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: The current code also fails for me (Windows PowerShell):
[output.txt](https://github.com/Lombiq/Utility-Scripts/files/9159378/output.txt)

Comment (sarahelsaig) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: If you run it on the dev branch you will get the same errors too. Also Utility-Scripts are still reloaded. It just fails with the modules in C:\Windows\system32\WindowsPowerShell\v1.0\Modules and C:\Program Files\WindowsPowerShell\Modules. I don't know if we can or should do anything about this, since trying to reload the built-in modules seem pointless anyway.

File: SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1 (hunk 5 above)
Comment (Piedone) [reply SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1:1]: Hmm, now it works.

File: Utilities/Reload-Module/Reload-Module.psm1 (hunk 14 above)
Comment (Piedone) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: I see, indeed, OK then.

Comment (sarahelsaig) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: I cleaned it up, so there will be no errors.

Comment (Piedone) [reply Utilities/Reload-Module/Reload-Module.psm1:16]: Thank you!

File: .gitmodules [hunk 15]
```diff
@@ -0,0 +1,4 @@
+[submodule "Analyzers"]
+	path = Analyzers
```
Comment (Piedone) [inline .gitmodules]: Why not `Lombiq.Analyzers.PowerShell`?

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by Piedone** [comment]
Would you like to review this once more, @BenedekFarkas? There are now some significant changes since your last review.

File: .gitmodules (hunk 15 above)
Comment (sarahelsaig) [reply .gitmodules]: To fit in with the other directories in the repo. Should I change it? If yes, what to? Lombiq.Analyzers.PowerShell or Analyzers/Lombiq.Analyzers.PowerShell

Comment (Piedone) [reply .gitmodules]: Let's do _tools/Lombiq.Analyzers.PowerShell_.

File: DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1 [hunk 16]
```diff
@@ -1,40 +1,55 @@
...
 
-        if ($pathItem -isnot [System.IO.FileInfo] -or $pathItem.Extension -ne ".csproj")
+        if ($pathItem -isnot [System.IO.FileInfo])
         {
             throw "The path `"$Path`" does not point to a Visual Studio project file!"
         }
 
-        $packageTextList = dotnet list $pathItem.FullName package
+        # In the future (once https://github.com/NuGet/Home/issues/7752 is resolved) we may be able to request this
+        # output in JSON or similar format. If so, this code should be replaced to use that and e.g. ConvertFrom-Json.
```
Comment (BenedekFarkas) [inline DotNetCLI/Get-VisualStudioProjectNuGetPackages/Get-VisualStudioProjectNuGetPackages.psm1:44]: The output format now is absolutely horrendous. :D

**Review by BenedekFarkas — APPROVED** [review]

File: .gitmodules (hunk 15 above)
Comment (sarahelsaig) [reply .gitmodules]: but that stands out from the rest :(

Comment (Piedone) [reply .gitmodules]: It also took me some time to get used to all the random lowercase-named folders in ASP.NET Core projects. It's still easier than having our own unique conventions :D.

Comment (sarahelsaig) [reply .gitmodules]: Ok, ugly it is!

Comment (Piedone) [reply .gitmodules]: image

Comment (sarahelsaig) [reply .gitmodules]: aaand of course it totally broke and sent me off on a complimentary wild goose chase! smh

Comment (Piedone) [reply .gitmodules]: Sending thoughts and prayers.

**Review by Piedone — APPROVED** [review]
